src/libraries/Microsoft.Extensions.Http/src/DefaultHttpClientFactory.cs

```diff
@@ -233,11 +245,44 @@ internal virtual void StopCleanupTimer()
         {
             lock (_cleanupTimerLock)
             {
-                _cleanupTimer!.Dispose();
+                _cleanupTimer?.Dispose();
                 _cleanupTimer = null;
             }
         }
 
+        // protected for tests
+        protected bool _disposed;
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+            _disposed = true;
+
+            StopCleanupTimer();
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~DefaultHttpClientFactory()
+        {
+            Dispose(false);
+        }
```

Comment:
Is a finalizer actually necessary?  I understand having IDisposable.Dispose promptly clean up, but if it doesn't get disposed, finalization isn't going to meaningfully help, and making it finalizable makes it measurably more expensive to construct, plus it ensures instances that go undisposed will survive to higher generations.

Reply:
Hmmm.... I believe the container _should_ always dispose IDisposable singletons...
And even if container itself isn't properly disposed, and end up being finalized before the factory, there's still a chance of the timer firing between container's and factory's finalizations (and creating ILogger there would throw exception). I guess, I can remove the finalizer and just rely on swallowing exceptions in that case...